### ISSUE TITLE
Option to filter-out entire ANSI escape sequences for `qrexec-client(-vm)` and/or `qvm-run`

### ISSUE THREAD

**Issue by alimirjamali** [QubesOS/qubes-issues#9482]
[How to file a helpful issue](https://www.qubes-os.org/doc/issue-tracking/)

### The problem you're addressing (if any)
The current existing code of [libqrexec-utils/do_replace_chars](https://github.com/QubesOS/qubes-core-qrexec/blob/main/libqrexec/replace.c) merely replaces escape characters with underline. This is problematic for the programs which do not have the option/configuration for avoiding [ANSI escape sequences](https://en.wikipedia.org/wiki/ANSI_escape_code#Fe_Escape_sequences). Most specifically text coloring and graphic rendition sequences. Just filtering the escape character and not the entire sequence renders the output hard to read/parse.

One of the examples is DNF5 which will be the default in Fedora 41 (discussed [here on forum](https://forum.qubes-os.org/t/qubes-os-r4-3-alpha-preview/27963/39))

### The solution you'd like
Either improving the existing `-t, -T, --filter-escape-chars*` options to filter-out the entire known sequences instead of just replacing the escape character with underline. Or introducing new options for proper filtering.

### The value to a user, and who that user might be
- Users will not require to learn/specify plain-text output option for individual programs (if applicable) and could use a generic option for all.
- Without this option, `qubes-dom0-update` output in R4.3 will be bad (the only existing alternative is to disable the progress bar).

### Completion criteria checklist

(This section is for developer use only. Please do not modify it.)

**Comment by marmarek** [comment]
The current option is meant for safety, not for data transformation. Remote side should not send any control sequences, but the option ensures even if it would do, it would not be effective. It's all to protect terminal emulator from potential attacks (there were many vulnerabilities like this in the past, and some are still applicable).
Technically, I'd also avoid variable-length sequence processing in the receiving side - there are many corner cases, like detecting where such sequence would end, handling sequences split across multiple data chunks etc. Getting it wrong may mean for example allowing some sequence that should be filtered otherwise. That's extra attack surface on the trusted side (especially since in this case it would apply to the part running in dom0).

The proper solution is for the remote side to not send those sequences. If the application in question (DNF5) doesn't support turning it off, it can be wrapped in a tool that does filtering on the remote side. Maybe even `sed` would suffice?

But also, what about (asking upstream to) adding an option to disable coloring? Or, stop wrapping the call in `script`, so that DNF doesn't have PTY, but a socket/pipe for stdin/out and shouldn't use ANSI sequences then. This will likely break real time download progress bar, though...

**Comment by alimirjamali** [comment]
> Maybe even `sed` would suffice?

Yes. Sed could do that. There are [ready to use Regex formulas](https://stackoverflow.com/a/14693789) for that.

> But also, what about (asking upstream to) adding an option to disable coloring?

I can work on that. But other programs suffer from the same symptom. Gentoo's emerge is one of them.

> Or, stop wrapping the call in `script`

This is what I am currently doing personally. Maybe this should be the default in R4.3? Most people are using GUI updater anyways and do not see progress bar at all.

**Comment by alimirjamali** [comment]
One more idea on how to make this safe and eliminate any possible extra attack surface. The filtering code could do its job based on Regex formulas and well known/well tested regex replace libraries. Finally the old code could replace any possible/remaining instances of escape codes with underlines.

**Comment by marmarek** [comment]
> This is what I am currently doing personally. Maybe this should be the default in R4.3? Most people are using GUI updater anyways and do not see progress bar at all.

I guess that's a short term solution.
The proper solution is to refactor dom0 update mechanism for a proper stdout/err signaling protocol (so, it reports machine-readable info that qubes-dom0-update or any frontend like qubes-update-gui for it can use for a proper progress bar). But that's obviously a lot more work. Hopefully we'll be there one day @piotrbartman :wink:

**Comment by alimirjamali** [comment]
BTW, I explored the possibility of using an ANSI to Gtk's Pango markup converter. So the output in GUI Updater could be in full color. But I rejected the idea as all available libraries are not a part of Python official libraries and it might introduce further attack surface.

Even though it is replatively simple to write one from scratch, I guess it is better to avoid the idea since Qubes is a security oriented project.

**Comment by alimirjamali** [comment]
Related:
https://github.com/rpm-software-management/dnf5/issues/839

**Comment by alimirjamali** [comment]
> If the application in question (DNF5) doesn't support turning it off, it can be wrapped in a tool that does filtering on the remote side.

My PR to DNF5 to re-implement the `--color=<always|never|auto>` flag has been pending review for over a month. It is a low priority issue for DNF5 team. I guess it would be better to write the remote side wrapper tool.

**Comment by alimirjamali** [comment]
The PR to reintroduce `--color` option for dnf5 is reviewed and merged. After it is released for Fedora 41 & 42, it should be safe to implement `--color=never` for `qubes-dom0-update --action=<search|info|...>`; so its output would be properly rendered.

> The proper solution is to refactor dom0 update mechanism for a proper stdout/err signaling protocol (so, it reports machine-readable info that qubes-dom0-update or any frontend like qubes-update-gui can use for a proper progress bar). But that's obviously a lot more work.

Yes. In addition to the above, the new `--color` option does not disable multi-progress lines which still sends ANSI control escape characters. But it might be too much work (with minor improvements).

> Still, I'd prefer this "advanced" filtering be done on the sending side. It could be even part of the qrexec (either built-in, or a helper wrapper that any service could use in their implementation).

Since I believe qubes project would benefit with a VM side advanced (ANSI escape sequence) filtering pipe which will have a broader use-cases than just the update mechanism, I will proceed with writing one. It should be able to do the job of filtering as well as replacing "script" wrapper utility. Considering it might have use cases beyond Qubes OS project, I guess it is better to select a generic name for it. Something like "script-sans-ANSI" .

**Comment by alimirjamali** [comment]
Reintroduction of `--color` for dnf5 is released to Fedora 41, 42 and Rawhide. It is now possible to include `--color=never` for `qubes-dom0-update` or `qubes-download-dom0-updates.sh`. PR will follow shortly
